src/particles/wakefields/WakeConvolution.cpp

```diff
@@ -0,0 +1,167 @@
+ *
+ * This file is part of ImpactX.
+ *
+ * Authors: Alex Bojanich, Chad Mitchell, Axel Huebl
+ * License: BSD-3-Clause-LBNL
+ */
+#include "WakeConvolution.H"
+#include "particles/ImpactXParticleContainer.H" //Includes all necessary AMReX headers
+#include "initialization/InitDistribution.H"
+
+#ifdef ImpactX_USE_FFT
+#include <ablastr/math/fft/AnyFFT.H>
+#endif
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <stdexcept>
+#include <vector>
+
+/*
+Wake Functions:
+
+Define the longitudinal and transverse resistive wall wake functions,
+the longitudinal CSR wake function, and their associated functions
+
+*/
+
+//Step Function
+double unit_step(double s)
+{
+    return s >= 0 ? 1.0 : 0.0; //If true return 1, else 0
+}
+
+//Alpha Function
+double alpha(double s)
+{
+    return 1.0 - alpha_1 * std::sqrt(s) - (1.0 - 2 * alpha_1) * s;
+}
+
+//Resistive Wall Wake Functions
+
+double w_t_rf(double s, double a, double g, double L)
+{
+    double s0 = (0.169 * std::pow(a, 1.79) * std::pow(g, 0.38)) / std::pow(L, 1.17);
+    double term = std::sqrt(std::abs(s) / s0) * std::exp(-std::sqrt(std::abs(s) / s0));
+    return (4 * Z0 * ablastr::constant::SI::c * s0 * unit_step(s)) / (M_PI * std::pow(a, 4)) * term;
+}
+
+double w_l_rf(double s, double a, double g, double L)
+{
+    double s00 = g * std::pow((a / (alpha(g / L) * L)), 2) / 8.0;
+    return (Z0 * ablastr::constant::SI::c * unit_step(s) * std::exp(-std::sqrt(std::abs(s) / s00))) / (M_PI * std::pow(a, 2));
+}
+
+//CSR Wake Function
+
+double w_l_csr(double s, amrex::ParticleReal R)
+{
+    double rc = std::pow(ablastr::constant::SI::q_e, 2) / (4 * M_PI * ablastr::constant::SI::ep0 * ablastr::constant::SI::m_e * std::pow(ablastr::constant::SI::c, 2));
+    double kappa = (2 * rc * ablastr::constant::SI::m_e * std::pow(ablastr::constant::SI::c, 2)) / std::pow(3, 1.0/3.0) / std::pow(R, 2.0/3.0);
+
+    return - (kappa * unit_step(s)) / std::pow(std::abs(s), 1.0/3.0);
+}
+
+//Convolution Function
+
+void convolve_fft(double* beam_profile, double* wake_func, int beam_profile_size, int wake_func_size, double delta_t, double* result, int padding_factor)
+{
+#ifdef ImpactX_USE_FFT
+    //Length of convolution result
+    int original_n = beam_profile_size + wake_func_size - 1; //Output size is n = 2N - 1, where N = size of signals 1,2
+
+    //Add padding factor to control amount of zero-padding
+    int n = static_cast<int>(original_n * padding_factor);
+
+    //Allocate memory for FFT inputs and outputs
+    using ablastr::math::anyfft::Complex;
+    double *in1 = (double*) malloc(sizeof(double) * n); //Allocate memory for 'n' real numbers for inputs and complex outputs
+    double *in2 = (double*) malloc(sizeof(double) * n);
+    Complex *out1 = (Complex*) malloc(sizeof(Complex) * n);
+    Complex *out2 = (Complex*) malloc(sizeof(Complex) * n);
+    Complex *conv_result = (Complex*) malloc(sizeof(Complex) * n);
+    double *out3 = (double*) malloc(sizeof(double) * n);
```

Comment:
We will as a last  step replace these allocations with `amrex::Gpu::DeviceVector<Real>` and copy in the data (host-device) from the preparatory calculations.